File: include/network/libqsslellipticcurve.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qsslellipticcurve.html
pub const qsslellipticcurve = struct {
    /// New constructs a new QSslEllipticCurve object.
    ///
    /// ``` other: QtC.QSslEllipticCurve ```
    pub fn New(other: ?*anyopaque) QtC.QSslEllipticCurve {
        return qtc.QSslEllipticCurve_new(@ptrCast(other));
    }

    /// New2 constructs a new QSslEllipticCurve object and invalidates the source QSslEllipticCurve object.
    ///
    /// ``` other: QtC.QSslEllipticCurve ```
    pub fn New2(other: ?*anyopaque) QtC.QSslEllipticCurve {
        return qtc.QSslEllipticCurve_new2(@ptrCast(other));
    }

    /// New3 constructs a new QSslEllipticCurve object.
    ///
    ///
    pub fn New3() QtC.QSslEllipticCurve {
        return qtc.QSslEllipticCurve_new3();
    }

    /// New4 constructs a new QSslEllipticCurve object.
    ///
    /// ``` param1: QtC.QSslEllipticCurve ```
    pub fn New4(param1: ?*anyopaque) QtC.QSslEllipticCurve {
        return qtc.QSslEllipticCurve_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: QtC.QSslEllipticCurve, other: QtC.QSslEllipticCurve ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslEllipticCurve_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: QtC.QSslEllipticCurve, other: QtC.QSslEllipticCurve ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QSslEllipticCurve_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#fromShortName)
    ///
    /// ``` name: []const u8 ```
    pub fn FromShortName(name: []const u8) QtC.QSslEllipticCurve {
        const name_str = qtc.struct_libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.QSslEllipticCurve_FromShortName(name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#fromLongName)
    ///
    /// ``` name: []const u8 ```
    pub fn FromLongName(name: []const u8) QtC.QSslEllipticCurve {
        const name_str = qtc.struct_libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        return qtc.QSslEllipticCurve_FromLongName(name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#shortName)
    ///
    /// ``` self: QtC.QSslEllipticCurve, allocator: std.mem.Allocator ```
    pub fn ShortName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslEllipticCurve_ShortName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslellipticcurve.ShortName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#longName)
    ///
    /// ``` self: QtC.QSslEllipticCurve, allocator: std.mem.Allocator ```
    pub fn LongName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSslEllipticCurve_LongName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qsslellipticcurve.LongName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#isValid)
    ///
    /// ``` self: QtC.QSslEllipticCurve ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return qtc.QSslEllipticCurve_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#isTlsNamedCurve)
    ///
    /// ``` self: QtC.QSslEllipticCurve ```
    pub fn IsTlsNamedCurve(self: ?*anyopaque) bool {
        return qtc.QSslEllipticCurve_IsTlsNamedCurve(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qsslellipticcurve.html#dtor.QSslEllipticCurve)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QSslEllipticCurve ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSslEllipticCurve_Delete(@ptrCast(self));
    }
};

File: src_original/network/libqsslellipticcurve.cpp
#include <QSslEllipticCurve>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qsslellipticcurve.h>
#include "libqsslellipticcurve.h"
#include "libqsslellipticcurve.hxx"

QSslEllipticCurve* QSslEllipticCurve_new(const QSslEllipticCurve* other) {
    return new QSslEllipticCurve(*other);
}

QSslEllipticCurve* QSslEllipticCurve_new2(QSslEllipticCurve* other) {
    return new QSslEllipticCurve(std::move(*other));
}

QSslEllipticCurve* QSslEllipticCurve_new3() {
    return new QSslEllipticCurve();
}

QSslEllipticCurve* QSslEllipticCurve_new4(const QSslEllipticCurve* param1) {
    return new QSslEllipticCurve(*param1);
}

void QSslEllipticCurve_CopyAssign(QSslEllipticCurve* self, QSslEllipticCurve* other) {
    *self = *other;
}

void QSslEllipticCurve_MoveAssign(QSslEllipticCurve* self, QSslEllipticCurve* other) {
    *self = std::move(*other);
}

QSslEllipticCurve* QSslEllipticCurve_FromShortName(const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return new QSslEllipticCurve(QSslEllipticCurve::fromShortName(name_QString));
}

QSslEllipticCurve* QSslEllipticCurve_FromLongName(const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    return new QSslEllipticCurve(QSslEllipticCurve::fromLongName(name_QString));
}

libqt_string QSslEllipticCurve_ShortName(const QSslEllipticCurve* self) {
    QString _ret = self->shortName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QSslEllipticCurve_LongName(const QSslEllipticCurve* self) {
    QString _ret = self->longName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool QSslEllipticCurve_IsValid(const QSslEllipticCurve* self) {
    return self->isValid();
}

bool QSslEllipticCurve_IsTlsNamedCurve(const QSslEllipticCurve* self) {
    return self->isTlsNamedCurve();
}

void QSslEllipticCurve_Delete(QSslEllipticCurve* self) {
    delete self;
}
